Water 1.0

QuintillusCFC opened the pull request:
Start differentiating land from water.  Human units can only move on land (sorry, no human boats yet).  Barbarians will build galleys if they are coastal, and galleys will only move on the coast, whereas barbarian warriors will stay on land.  Barbarian camps will also always be on land instead of sometimes being on water.

Also has a few stylistic changes as identified in part with Rider.  Some of these are whitespace fixes, so if the two halves of the comparison look the same, that's probably what's going on.

Finally, changed the wait time for AI turns to 1 second from 2.  Now that there is barbarian activity, we don't really need a whole two second wait between turns.

QuintillusCFC committed:
Add a check for whether a tile is water.  Barbarians no longer spawn on water.  Although they will still walk on water.

QuintillusCFC committed:
Teach barbarians that they should stay on land, not on water.

Mainly this consists of adding a method on Tile that returns a random land neighbor, not a totally random neighbor.

QuintillusCFC committed:
Fix an area with mixed-indentation lines that I ran across.

QuintillusCFC committed:
Some refactorings around barbarians.  Primarily adding the barbarianCamps list to the map so we don't have to iterate through all the tiles.  But also some stylistic ones that Rider helped identify.

QuintillusCFC committed:
Teach the barbarians how to build galleys, which will only travel on the coast.  Only coastal barbarians will build galleys.

QuintillusCFC committed:
Player-controlled units can no longer walk on water.

maxpetul commented:
Instead of creating a new Random object every time you need a random number, you should instead use the Random object stored in GameData (named "rng"). The problem with creating a new object each time is that it gets a new seed based on the system clock which will prevent us from implementing the preserve-random-seed option and more importantly will prevent us from keeping the stream of "random" numbers consistent for all players in a multiplayer game.

maxpetul commented on C7GameData/Tile.cs:
```diff
@@ -74,6 +75,22 @@ public override string ToString()
 		}
+
+		public Tile RandomLandNeighbor() {
+			List<Tile> landNeighbors = neighbors.Values.ToList().Where(tile => !tile.baseTerrainType.isWater()).ToList();
```
It's a minor thing but you don't need the first call to ToList here or the corresponding one in RandomCoastNeighbor. "Where" works on any IEnumerable, which is implemented by the values collection, so there's no need to gather the values into a list.

QuintillusCFC committed:
Don't create random number generators everywhere; re-use the one in GameData.cs.

This will allow the 'preserve random seed' option to work, and lessen the chance of save scumming.

Note that in the "Quick Start" option, it's deserializing the game state from JSON, so we have a hack to set an RNG there since our JSON doesn't currently have an RNG.

QuintillusCFC commented:
Good call, I had forgotten about Preserve Random Seed.  Also makes sense that we'll want it to be consistent in MP, should we get to that point.

I've pushed some changes for that; one note is that because our "Quick Start" option loads from the JSON that currently lacks an RNG, there's a hack to set the RNG once, in the load, if it's null.  Once we get a more built-out standard rule set/random game generation that shouldn't be necessary, but I don't want to scope creep this PR to incorporate that.

QuintillusCFC committed:
Remove unnecessary ToList() before the Where clause.
